[ADD] Modify sale delivery address

StefanRijnhart opened the pull request:
# 
# Modify sale delivery address

For sales orders that have not been shipped completely, this module allows the
user to change the delivery address on the sales order for the remaining
deliveries.

StefanRijnhart committed:
[ADD] Modify sale delivery address

pedrobaeza commented:
Why don't simply change the partner on the remaining picking?

StefanRijnhart commented:
I'm following the way that the address is propagated in standard Odoo, to prevent any unforeseen inconsistencies from occurring.

pedrobaeza commented on sale_delivery_address_modify/models/sale_order.py:
```diff
@@ -0,0 +1,55 @@
+            for order in self:
+                if not order.procurement_group_id or order.shipped:
+                    continue
+                for procurement in self.env['procurement.order'].search(
```
You don't need to search for procurements. Search only pickings with sale_id field = order.id

StefanRijnhart replied:
Shouldn't that be equivalent?

pedrobaeza replied:
Well, you save 1 search with my expression

StefanRijnhart replied:
Tested it, but it does not work in the following situation: create a sale order and confirm. Update the delivery address. This updates the delivery address on the picking. Cancel the picking. On the sale order, recreate the picking. The picking has the old delivery address because it is taken from the procurement, not the sale order.

pedrobaeza replied:
OK, that flow is enough to justify this module, although this inline comment refers to the search expression, not the general concern. You can add this rationale in the module description.

StefanRijnhart replied:
OK, then I did not understand what you meant. Description updated.

pedrobaeza replied:
I'm talking to use this code instead of the 2 searches:

``` python
            pickings = self.env['stock.picking'].search(
                    [('sale_id', '=', order.id), ('state', '!=', 'done')])
```

StefanRijnhart replied:
sale_id is a non searchable function field in 8.0. I could use

```
pickings = self.env['stock.picking'].search(
    [('id', 'in', (order.picking_ids.ids), ('state', '!=', 'done')])
```

Where picking_ids is a function field that triggers the same query on pickings that I am curently using, through the procurement group. Would that be an improvement?

pedrobaeza replied:
Ah, OK, I didn't know about the not searchable issue. What you can do then is:

``` python
pickings = order.picking.ids.filtered(lambda x: x.state != 'done')
```

More pythonic and readable, and if one day this field is stored, we will automatically get the improvement.

StefanRijnhart replied:
OK, done.

guewen commented:
> Why don't simply change the partner on the remaining picking?

Separation of concerns. I guess that's because the salespeople that need to change the address (for instance after a call of their customer) are not the same people that work on the delivery orders. Also you are sure not to forget a picking.

:+1: 

EDIT: @pedrobaeza I think I misunderstood your question, forget my answer ;-)

pedrobaeza commented:
OK, understood

StefanRijnhart commented:
Seems that the INCLUDE/EXCLUDE config messes up coverage. Is there a solution to that?

pedrobaeza commented:
Not for now. See https://github.com/OCA/maintainer-quality-tools/issues/249

Ideal for me is not to have INCLUDE/EXCLUDE.

pedrobaeza commented:
@guewen, analysing again your answer, I'm still reluctant: a salesman can access the delivery(ies) order(s) via a button in the sales order itself, so no need to access warehouse menu. And he/she can change the date on the non-done picking (it will be always only one).

StefanRijnhart committed:
[UPD] Description

StefanRijnhart committed:
[IMP] Filter out unprocessed pickings more pythonically

pedrobaeza commented on sale_delivery_address_modify/models/sale_order.py:
```diff
@@ -0,0 +1,54 @@
+                    lambda picking: picking.state != 'done')
+                pickings.write(
+                    {'partner_id': order.partner_shipping_id.id})
+                self.env['stock.move'].search(
```
Another optional way to do this is (not more efficient, but more readable):

``` python
pickings.mapped('move_ids').write(...
```

StefanRijnhart replied:
Done.

pedrobaeza commented:
Thanks :+1:

guewen commented:
Travis is red due to 

```
sale_delivery_address_modify/__openerp__.py:20: [C8102(manifest-required-key), ] Missing required key "license" in manifest file
```

StefanRijnhart commented:
Let me restart as https://github.com/OCA/maintainer-quality-tools/pull/315 has been merged

atchuthan commented on sale_delivery_address_modify/models/sale_order.py:
```diff
@@ -0,0 +1,53 @@
+                        raise UserError(_(
+                            'The delivery address can only be changed if '
+                            'the related customer location remains the same.'))
+                        procurement.write({'partner_dest_id':
```
after error raises with UserError, the next lines won't be executed

atchuthan commented:
Other than a dead block, code LGTM :+1:

StefanRijnhart committed:
[FIX] indendation level

StefanRijnhart commented:
Thanks @atchuthan, fixed in https://github.com/OCA/sale-workflow/pull/233/commits/7f9c3f4b7da2e5eb546d621385694f5d0fb07f94. Rebased to see what travis comes up with now.

StefanRijnhart committed:
[RFR] Use old api for copy operation as per bug

StefanRijnhart commented:
Travis is green as tests now succeed after working around a bug in Odoo. Runbot is still failing for reasons unrelated to this PR (@gurneyalex, can we have bank-statement-reconcile as a dependency of this repo in Runbot?)

pedrobaeza commented:
I have added the dependency and rebuild the instance.

StefanRijnhart commented:
Thanks @pedrobaeza, all green now!

pedrobaeza commented on sale_delivery_address_modify/i18n/sale_delivery_address_modify.pot:
```diff
@@ -0,0 +1,28 @@
+# Translation of Odoo Server.
```
Remove the pot file, as this is no longer needed and can confuse people if it's not up to date.

pedrobaeza commented on sale_delivery_address_modify/models/sale_order.py:
```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
+##############################################################################
```
Can you please use short headers like in https://github.com/OCA/maintainer-tools/blob/master/template/module/__init__.py

StefanRijnhart committed:
[DEL] Pot file; [RFR] License

StefanRijnhart commented:
Sorry for the delay, POT file is now removed and new style license block is applied.

gurneyalex commented:
:+1:
